JavaScript/Next-js.md

```diff
@@ -0,0 +1,221 @@
+## Introduction
+
+Next.js supports some very cool features which can be used on different kind of projects. For example, some projects with public pages couldn't use create-react-app because it doesn't support server-side rendering, which makes SEO much harder. So in projects that require SSR, the different stack must be used that would probably include something like express on top of node, and custom webpack and babel configuration.
+
+Also, any customization of webpack or babel inside CRA requires ejecting, which is the one-way operation, and from that point, you are responsible for maintaining configuration files.
+
+Soon every project would probably introduce something new so we would end up with many different projects very hard to maintain.
+Next.js solves above-stated problems and allows us to use a single stack for many different projects. Plus next.js introduces some handy features, which comes out of the box:
+
+- Code splitting
+- SSR
+- Page prefetching
+- Dynamic import of modules
+- Built-in Zero-Config TypeScript Support
+- Integrated client-side routing
+- Custom server with custom express routes
+- static HTML export
+
+## POC
+
+We made POC which used some of the above features. We wanted to build an app which could be exported to static Html files (wouldn't need any server processing) and would contain modals which we could access via link.
+
+We used this requirement since many of our projects use modals, and it proved to be very useful to access them directly by adding query params to url.
+
+### App structure
+
+```
+...
+├─ components
+⎮   ├─ higherOrderComponents
+⎮   ⎮   └─ WithAuth.tsx
+⎮   ⎮   └─ WithModal.tsx
+⎮   ├─ layouts
+⎮   ⎮   └─ Layout.tsx
+⎮   ├─ modals
+|   |   └─ index.tsx
+|   |   └─ Newsletter.tsx
+|   └─ LoginRegisterForm.tsx
+├─ pages
+|   ├─ index.tsx
+⎮   ├─ login.tsx
+⎮   ├─ register.tsx
+|   └─ with-hoc-modal.tsx
+├─ services
+|   └─ auth.ts
+...
+```
+
+`login.tsx` and `register.tsx` are public pages whose functionality is self-explanatory. Once login is successful user is redirected to the index page which acts as the home page.
+
+### Layouts
+
+`Layout.tsx` is an example of a layout page for all private pages. Here is an example of how we can extend `Head` section for adding global fonts, but more importantly, we use layout page with higher order component `WithAuth` which is responsible for authentification.
```

Comment:
Layout should take care of rendering ... well ... layouts xD ... navigation, content, footer and some common global styles... and we want it to work on ssr, on client only, with logged in user or logged out user...